ttnn/cpp/ttnn/operations/pool/generic/pool2d_utils.cpp

```diff
@@ -0,0 +1,477 @@
+                input_tensor_layout == Layout::ROW_MAJOR) {
+                log_debug(tt::LogOp, "Auto shard, enable shallow conv");
+                // height sharded, non matmul conv, with input channels < 32, and default setting for
+                // input_channels_alignment
+                // Currently data-movement ops have too many restrictions to support shallow convs with tiled input.
+                conv_config.input_channels_alignment = 8;
+            } else if (shard_layout != TensorMemoryLayout::HEIGHT_SHARDED) {
+                conv_config.input_channels_alignment = tt::constants::TILE_WIDTH;
+            }
+
+            // Set act_block_h_override to min value to
+            // be conservative with L1 memory usage.
+            uint32_t act_block_h_override = tt::constants::TILE_HEIGHT;
+        }
+
+        const uint32_t in_channels_padded = ccl::cmd::round_up(channels, conv_config.input_channels_alignment);
+        const uint32_t output_channels_padded = ccl::cmd::round_up(channels, tt::constants::TILE_WIDTH);
+        // Note: These are not exact shapes for weights as prepare_conv_weights will pad the weights depending on the
+        // conv2d params, but these are good enough for L1 usage estimation.
+        // const ttnn::Shape weights_shape(
+        //     {1, 1, in_channels_padded * kernel_size[0] * kernel_size[1], output_channels_padded});
+
+        const sliding_window::ParallelConfig input_parallel_config = conv::determine_parallel_config(
+            shard_layout,
+            batch_size,
+            channels,
+            output_height,
+            output_width,
+            channels,
+            compute_grid_size,
+            shard_orientation,
+            false,
+            false);
+
+        const sliding_window::ParallelConfig output_parallel_config =
+            conv::determine_output_parallel_config(input_parallel_config, compute_grid_size, channels, false);
+
+        auto [opt_conv_op_parallel_config, opt_conv_op_block_config, conv_out_memory_config] = conv::get_conv_configs(
+            conv_config,
+            compute_config,
+            input_parallel_config,
+            output_parallel_config,
+            channels,
+            channels,
+            batch_size,
+            output_height,
+            output_width,
+            kernel_size,
+            compute_grid_size);
+
+        if (conv_config.act_block_w_div == 1 && conv_config.shard_layout == TensorMemoryLayout::WIDTH_SHARDED) {
+            uint32_t width_sharded_num_cores = input_parallel_config.grid.num_cores();
+            // Set act_block_w_div to max value to
+            // be conservative with L1 memory usage.
+            // act_block_w_div == 1 is currently the default value.
+            conv_config.act_block_w_div = tt::div_up(channels, width_sharded_num_cores * tt::constants::TILE_WIDTH);
+        }
+
+        conv::conv_op_l1_usage l1_usage = pool::calculate_L1_usage(
```

Comment:
In conv2d we validate that l1 usage calculation is correct by asserting that op really took that much memory in runtime.
This ensures that l1 calculation will stay accurate over time.
You can take a look of how it's done here
https://github.com/tenstorrent/tt-metal/blob/1db8852d08b1e1ad0aabbd655f8ddc414ef2c0a9/ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op.cpp#L335